open_spiel/games/2048.cc

```diff
@@ -0,0 +1,378 @@
+  PrepareTiles();
+  for (int r : traversals[0]) {
+    for (int c : traversals[1]) {
+      int tile = GetCellContent(r, c);
+      if (tile > 0) {
+        bool moved = false;
+        std::array<Coordinate, 2> positions = FindFarthestPosition(r, c, action);
+        Coordinate farthest_pos = positions[0];
+        Coordinate next_pos = positions[1];
+        int next_cell = GetCellContent(next_pos.row, next_pos.column);
+        if (next_cell > 0 && next_cell == tile
+            && !BoardAt(next_pos).is_merged) {
+          int merged = tile * 2;
+          action_score_ += merged;
+          SetBoard(next_pos.row, next_pos.column, Tile(merged, true));
+          moved = true;
+        } else if (farthest_pos.row != r || farthest_pos.column != c){
+          SetBoard(farthest_pos.row, farthest_pos.column, Tile(tile, false));
+          moved = true;
+        }
+        if (moved) {
+          SetBoard(r, c, Tile(0, false));
+          current_player_ = kChancePlayerId;
+        }        
+      }
+    }
+  }  
+  total_score_ += action_score_;
+}
+
+std::string TwentyFortyEightState::ActionToString(Player player,
+                                          Action action_id) const {
+  if (IsChanceNode()) {
+    if (action_id == kNoCellAvailableAction) {
+      return "No Cell Available";
+    }
+    ChanceAction chance_action = SpielActionToChanceAction(action_id);
+    return absl::StrCat(chance_action.is_four ? 4 : 2, 
+        " added to row ", chance_action.row + 1,
+        ", column ", chance_action.column + 1);
+  }
+  switch (action_id) {
+    case kMoveUp:
+      return "Up";
+    case kMoveRight:
+      return "Right";
+    case kMoveDown:
+      return "Down";
+    case kMoveLeft:
+      return "Left";
+    default:
+      return "Invalid action";
+  }  
+}
+
+int TwentyFortyEightState::AvailableCellCount() const {
+  int count = 0;
+  for (int r = 0; r < kRows; r++) {
+    for (int c = 0; c < kColumns; c++) {
+      if (BoardAt(r, c).value == 0) {
+        count++;
+      }
+    }
+  }
+  return count;
+}
+
+ActionsAndProbs TwentyFortyEightState::ChanceOutcomes() const {  
+  int count = AvailableCellCount();
+  if (count == 0) {
+    return {{kNoCellAvailableAction, 1.0}};    
+  }
+  ActionsAndProbs action_and_probs;
+  action_and_probs.reserve(count * 2);
+  for (int r = 0; r < kRows; r++) {
+    for (int c = 0; c < kColumns; c++) {
+      if (BoardAt(r, c).value == 0) {
+        // 2 appearing randomly on the board should be 9 times as likely as a 4
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, false)), .9 / count);
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, true)), .1 / count);
+      }      
+    }
+  }  
+  return action_and_probs;
+}
+
+std::vector<Action> TwentyFortyEightState::LegalActions() const {
+  if (IsTerminal()) {
+    return {};
+  }
+  if (IsChanceNode()) {
+    return LegalChanceOutcomes();
```

Comment:
LegalChanceOutcomes doesn't seem to be defined anywhere. Is this ChanceOutcomes()?

Reply:
I was referencing backgammon.cc for writing this.
As I understand, LegalChanceOutcomes() is defined in spiel.h and it uses the output of ChanceOutcomes() to return vector<Action>